### PR TITLE
Convolutions tasks divided into different files

### PR DESCRIPTION
**Pull request by AryanNanda17**
This issue involves refactoring the provided convolution code into three separate files, each focusing on a specific aspect of convolution operations, and creating a separate folder for benchmarking the results of each file.
Resolved issue #120

### CODE REVIEW

**Comment by AryanNanda17** [comment]
@PritK99, please checkout the PR and let me know if any change is required.

**Comment by PritK99** [comment]
Why are folders such as ```/src``` and ```/include``` a part of ```/benchmark``` directory?  The idea was to just create a new folder ```/benchmark``` to deal with the time features.

**Comment by AryanNanda17** [comment]
In the benchmarks folder time taken by each convolution is also being printed.  In order to Benchmark(time taken by convolution operation), we would have to perform convolution, so the folders ```/src``` and ```/include``` are important.

**Comment by advait-0** [comment]
@AryanNanda17 please add the usage instructions to run each file in the sub-folder readme.

**Comment by AryanNanda17** [comment]
@advait-0, I made the required changes please check it out.

**Comment by aPR0T0** [comment]
Hello @AryanNanda17, Please add benchmarks to each file or remove them from the separable convolutions as they all should have a common standard. The point of the benchmark here is to understand the reason behind the use of separable convolutions

**Review by aPR0T0 — CHANGES_REQUESTED** [review]
Please add / remove benchmarks

**Comment by AryanNanda17** [comment]
> Hello @AryanNanda17, Please add benchmarks to each file or remove them from the separable convolutions as they all should have a common standard. The point of the benchmark here is to understand the reason behind the use of separable convolutions

Hello @aPR0T0, could you please elaborate on this?
Because from what I understand issue #120 aims to make the code easier. So, I have separated all three approaches of convolutions in our case into three different files without benchmarks. For benchmarks, I have created a separate folder ```/benchmarks```
Screenshot 2024-02-24 at 12 19 07 AM
The benchmarks folder benchmarks all three types of convolutions in our case!

**Comment by AryanNanda17** [comment]
> Sure @AryanNanda17 , makes sense then. Great work!
> 
> Just a small change, add benchmark ka example execution in readme

Thanks!

**Comment by AryanNanda17** [comment]
@aPR0T0, I added the results of Benchmarks in Readme. 
Please check it out.

File: 4_cv_basics/4_convolutions_filtering/benchmarks/README.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+## A suboptimal implementation
+
+We have spun up a suboptimal implementation of the convolution operation for
+demonstration purposes.  
+To build and run it, execute the following commands from the root of this
+repository.
+
+```bash
+cd 4_cv_basics/4_convolutions_filtering      # Enter the project directory
+make SRC=main.cpp link=src/convolution.cpp   # Use make to build the project
```
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/benchmarks/README.md:203]: Can’t find a `main.cpp` maybe this should be `native.cpp`

File: 4_cv_basics/4_convolutions_filtering/benchmarks/Makefile
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/benchmarks/Makefile:1]: Looks good to me

File: 4_cv_basics/4_convolutions_filtering/convolutionUsingOpenCV.cpp
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/convolutionUsingOpenCV.cpp:1]: Looks Good to me

File: 4_cv_basics/4_convolutions_filtering/naive.cpp
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/naive.cpp:1]: looks good to me!

File: 4_cv_basics/4_convolutions_filtering/separableConvolutions.cpp
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/separableConvolutions.cpp:1]: Looks good to me!

**Review by SuperChamp234 — CHANGES_REQUESTED** [review]
Screenshot 2024-02-24 at 9 05 03 AM


Why does the benchmarking folder contain the same files from the parent folder?

Ideally, the benchmarking folder should just have a file that will use the convolutions demonstrated in it’s parent folder and show the difference. Different files for each of the convolutions is not necessary.

**Comment by SuperChamp234** [comment]
Please make the change of removing unnecessary files and merge the benchmarking code into one file @AryanNanda17. Good job so far!

**Comment by AryanNanda17** [comment]
> Please make the change of removing unnecessary files and merge the benchmarking code into one file @AryanNanda17. 

Hello @SuperChamp234!
- In issue #120 , @PritK99 suggested to make three separate files for each type of convolution and create a separate folder to benchmark the results of **each convolution file**. If we merge all three convolution files into one code, the code would become hard as before. But our goal here is to enhance code modularity, readability, and maintainability by isolating different functionalities into distinct files.

**Comment by aPR0T0** [comment]
> > Please make the change of removing unnecessary files and merge the benchmarking code into one file @AryanNanda17.
> 
> Hello @SuperChamp234!
> 
>     * In issue [Refactoring Convolution Filtering #120](https://github.com/SRA-VJTI/Pixels_Seminar/issues/120) , @PritK99 suggested to make three separate files for each type of convolution and create a separate folder to benchmark the results of **each convolution file**. If we merge all three convolution files into one code, the code would become hard as before. But our goal here is to enhance code modularity, readability, and maintainability by isolating different functionalities into distinct files.

Hello @AryanNanda17 , that's right we need to simplify the code

Just restructure the directories as ->
1. benchmark
2. include
3. src
4. assets
other files

and inside benchmark just keep cpp files, this way the folder remains generic

**Comment by AryanNanda17** [comment]
@aPR0T0, I made the changes. Please check it out.

**Review by aPR0T0 — APPROVED** [review]
LGTM!
Great work @AryanNanda17 !

File: 4_cv_basics/4_convolutions_filtering/README.md
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/README.md:1]: looks good to me!

File: 4_cv_basics/4_convolutions_filtering/benchmarks/convolutionUsingOpenCV.cpp
Comment (SuperChamp234) [inline 4_cv_basics/4_convolutions_filtering/benchmarks/convolutionUsingOpenCV.cpp:1]: looks good to me!

**Review by SuperChamp234 — APPROVED** [review]

**Pull request merged**